roaring/roaring.go

```diff
@@ -395,32 +399,299 @@ func (b *Bitmap) Intersect(other *Bitmap) *Bitmap {
+//        900 + 3500 > ArrayMaxSize so we will need to upgrade to a bitset container which
+//        we will have to allocate, and then we will have to perform two unions into the
+//        new bitmap container: one for the array container generated in the previous step,
+//        and one for the bitset container in bitmap C.
+//
+// Approach taken by this function:
+//
+//     1. Detect that bitmaps A, B, and C all have containers for key 0.
+//     2. Estimate the resulting cardinality of the union of all their containers to be
+//        400 + 500 + 3500 > ArrayMaxSize and decide upfront to use a bitset for the target
+//        container.
+//     3. Union the containers from bitmaps A, B, and C into the new bitset container directly
+//        using fast bitwise operations.
+//
+// In the naive approach, we had to allocate two containers, whereas in the optimized approach
+// we only had to allocate one container, and we also had to perform less union operations. This
+// example is simplistic, but the impact in terms of CPU cycles and memory allocations achieved
+// by using the optimized alogorithm when unioning many large bitmaps can be huge.
+//
+// An additional optimization that this function makes is that it recognizes that even when
+// CPU support is present, performing the popcount() operation isn't free. Imagine a scenario
+// where 10 bitset containers are being unioned together one after the next. If every
+// bitset<->bitset union operation needs to keep the containers cardinality up to date, then
+// the algorithm will waste a lot of time performing intermediary popcount() operations that
+// will immediately be invalidated by the next union operation. As a result, we allow the cardinality
+// of containers to degrade when we perform the in-place union operations, and then when the algorithm
+// completes we "repair" all the containers by performing the popcount() operation one time. This means
+// that we only ever have to do O(1) popcount operations per container instead of O(n) where n is the
+// number of containers with the same key that are being unioned together.
+//
+// The algorithm works by iterating through all of the containers in all of the bitmaps concurrently.
+// At every "tick" of the outermost loop, we increment our pointer into the bitmaps list of containers
+// by 1 (if we haven't reached the end of the containers for that bitmap.)
+//
+// We then loop through all of the "current" values(containers) for all of the bitmaps
+// and for each container with a specific key that we encounter, we scan forward to see if any of the
+// other bitmaps have a container for the same key. If so, we calculate some summary statistics and
+// then use that information to make a decision about how to union all of the containers with the same
+// key together, perform the union, mark the unioned containers as "handled" and then move on to the next
+// batch of containers that share the same key.
```

Comment:
I was trying to think why "handled" exists, and the answer is, we don't have any guarantee or even strong expectation that every container-key exists in even one bitmap. But I think it might be possible to streamline this logic.

Consider a heap of bitmaps, with the ordering function being "next container key", lowest-first. You pick the lowest thing off the heap, grab its container, iterate it, and push it back into the heap. As long as the top of the heap has the same key as your current container, do the same thing to that. When you run out of things with the same key, you have a pool of containers with the same key, and the heap contains the next likely key...

But that'll be horrendously expensive in the probably-common case where most bitmaps have most containers.

It does seem, though, like somehow arranging to process by "lowest container key first" might reduce bookkeeping; you wouldn't have to track "handled" if you weren't possibly handling a key that's higher than some bitmap's "current" key. And intuitively, for at least some of the formats, I think you'll get better performance iterating through them strictly in order, rather than suddenly grabbing something from far later in a bitmap and then jumping back.

Reply:
Yeah I see what you're saying. If you think about it though the `handledIter` slice thing is basically my way of doing it. If I tucked the `handled` logic into the `handledIter` datastructure itself, I could make it expose a separate iterator that provided the semantics of "lowest container key" iteration. I actually tried doing that in one implementation and I didn't like how it made the code read because its kind of a leaky abstraction and things get messy quickly.

Ultimately though we have to deal with the data in the format that it is. Even small amounts of allocations or restructuring will degrade the performance and memory utilization significantly. The reason this implementation is so fast is because it doesn't restructure the data at all, and in the optimistic case, I don't even have to allocate the `handledIter` slice.

Its not my favorite part of this code either, but I don't see an easy way to improve this that won't introduce a bunch of allocations or significantly complicate things.